src/main/telemetry/mavlink.c

```diff
@@ -1202,24 +1232,23 @@ void handleMAVLinkTelemetry(timeUs_t currentTimeUs)
         return;
     }
 
-    // If we did serve data on incoming request - skip next scheduled messages batch to avoid link clogging
-    if (processMAVLinkIncomingTelemetry()) {
-        incomingRequestServed = true;
+    // Process incoming MAVLink - ignore the return indicating whether or not a message was processed
+    // Very few telemetry links are dynamic uplink/downlink split so uplink telemetry shouldn't reduce downlink bandwidth availability
+    processMAVLinkIncomingTelemetry();
+
+    // Determine whether to send telemetry back
+    bool shouldSendTelemetry = false;
+    if (txbuff_valid) {
+        // Use flow control if available
+        shouldSendTelemetry = txbuff_free >= 33;
+    } else {
+        // If not, use blind frame pacing
+        shouldSendTelemetry = (currentTimeUs - lastMavlinkMessage) >= TELEMETRY_MAVLINK_DELAY;
     }
 
-    if ((currentTimeUs - lastMavlinkMessage) >= TELEMETRY_MAVLINK_DELAY) {
-        // Only process scheduled data if we didn't serve any incoming request this cycle
-        if (!incomingRequestServed ||
-            (
-                 (rxConfig()->receiverType == RX_TYPE_SERIAL) &&
-                 (rxConfig()->serialrx_provider == SERIALRX_MAVLINK) &&
-                 !tristateWithDefaultOnIsActive(rxConfig()->halfDuplex)
-            )
-        ) {
-            processMAVLinkTelemetry(currentTimeUs);
-        }
+    if (shouldSendTelemetry) {
```

Comment:
I think the wait before sending makes sense for half duplex links. You want to make sure the link is not busy when you start to transmit. A short wait to ensure the is silent is a common strategy for this. ELRS mavlink rc is delegating that de-conflicting down the line, so we can treat it as full duplex with flow control as we won't have collisions on the uart itself.

So something along the lines of:

if(!isHalfDuplex || (!incommingRequestServed)) would keep the current half duplex behavior, but we need to tell INAV to treat mavlinkrc as full duplex.

The check should probably be expanded to check for serialrx_halfduplex and telemetry_halfduplex.

I have to dig in the code a big more, and validate who owns the uart. The tristateWithDefaultOnIsActive  check on rxConfig is probably wrong, if telemetry_halfduplex is set to OFF. A workaround would be to set both to full duplex for validation.

Reply:
This check probably makes more sense for half duplex on vs off.
```
static bool isMAVLinkTelemetryHalfDuplex(void) {
    return  (rxConfig()->receiverType == RX_TYPE_SERIAL && rxConfig()->serialrx_provider == SERIALRX_MAVLINK && tristateWithDefaultOffIsActive(rxConfig()->halfDuplex))
            || telemetryConfig()->halfDuplex;
}
```